optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/solver/EmployeeRosteringConstraintProvider.java

```diff
@@ -0,0 +1,332 @@
+                .penalizeConfigurable(constraintName);
+    }
+
+    Constraint lowRiskEmployeeAssignedToCovidWard(ConstraintFactory constraintFactory) {
+        return riskEmployeeAssignedToCovidWard(constraintFactory, LOW,
+                CONSTRAINT_COVID_LOW_RISK_EMPLOYEE);
+    }
+
+    Constraint moderateRiskEmployeeAssignedToCovidWard(ConstraintFactory constraintFactory) {
+        return riskEmployeeAssignedToCovidWard(constraintFactory, MODERATE,
+                CONSTRAINT_COVID_MODERATE_RISK_EMPLOYEE);
+    }
+
+    Constraint highRiskEmployeeAssignedToCovidWard(ConstraintFactory constraintFactory) {
+        return riskEmployeeAssignedToCovidWard(constraintFactory, HIGH,
+                CONSTRAINT_COVID_HIGH_RISK_EMPLOYEE);
+    }
+
+    Constraint extremeRiskEmployeeAssignedToCovidWard(ConstraintFactory constraintFactory) {
+        return riskEmployeeAssignedToCovidWard(constraintFactory, EXTREME,
+                CONSTRAINT_COVID_EXTREME_RISK_EMPLOYEE);
+    }
+
+    Constraint inoculatedEmployeeOutsideCovidWard(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
+                .filter(shift -> !shift.getSpot().isCovidWard())
+                .filter(shift -> shift.getEmployee().getCovidRiskType() == INOCULATED)
+                .penalizeConfigurable(CONSTRAINT_COVID_INOCULATED_EMPLOYEE_OUTSIDE_COVID_WARD);
+    }
+
+    Constraint uniformDistributionOfInoculatedHours(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
+                .filter(shift -> shift.getEmployee().getCovidRiskType() == INOCULATED)
+                .groupBy(new DefaultUniConstraintCollector<Shift, Map<OffsetDateTime, Integer>, Collection<Integer>>(
+                        () -> new HashMap<>(0),
+                        (container, shift) -> {
+                            shift.increaseHourlyCounts(container);
+                            return () -> shift.decreaseHourlyCounts(container);
+                        },
+                        Map::values
+                ))
+                .penalizeConfigurableLong(CONSTRAINT_COVID_UNIFORM_DISTRIBUTION_OF_INOCULATED_HOURS,
+                        Shift::calculateLoad);
+    }
+
+    Constraint maximizeInoculatedHours(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
+                .filter(shift -> shift.getEmployee().getCovidRiskType() == INOCULATED)
+                .rewardConfigurableLong(CONSTRAINT_COVID_MAXIMIZE_INOCULATED_HOURS, Shift::getLengthInHours);
+    }
+
+    Constraint migrationBetweenCovidAndNonCovidWards(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
```

Comment:
I wonder if this can be improved using a joiner:

```
constraintFactory.from(Shift.class)
    .factFilter(s -> s.getSpot().isCovidWard) // It be nice if we have a separate filter for facts that do not change,
     // so that this filter does not get rerun on a Shift when it changes
    .join(Shift.class, Joiners.notEquals(s -> s.getSpot().isCovidWard()), Joiners.equals(Shift::getEmployee), Joiners.greaterThanOrEqual(Shift::getEndDateTime, Shift::getStartDateTime))
    .ifNotExists(Shift.class, /* same code */)
    .penalizeConfigurable(...)
```

Reply:
It would be possible, only `notEqual()` joiner currently does not exist. I have a JIRA for that in current sprint, but it will likely have to take a back seat due to this COVID work.

Reply:
Do booleans implement `Comparable` (i.e. is `false < true` or vice-versa)?

Reply:
They do, but I don't consider it very intuitive to rely on that behavior.